modules/gapi/test/gapi_transform_tests.cpp

```diff
@@ -103,81 +116,74 @@ TEST(KernelPackageTransform, Combine)
 
-TEST(KernelPackageTransform, Substitute)
+TEST(KernelPackageTransform, gmat_gsc_in_gmat_out)
 {
-    auto tr = gmat2_in_gmat3_out::transformation();
-    GMat a, b;
-    auto subst = tr.substitute({cv::GArg(a), cv::GArg(b)});
+    auto tr = gmat_gsc_garray_in_gmat2_out::transformation();
 
-    EXPECT_EQ(3u, subst.size());
-    for (const auto& s : subst)
-    {
-        EXPECT_NO_THROW(s.get<GMat>());
-    }
-}
-
-template <typename Transformation, typename InType, typename OutType>
-static void transformTest()
-{
-    auto tr = Transformation::transformation();
-    InType in;
-    auto pattern = tr.pattern({cv::GArg(in)});
-    auto subst = tr.substitute({cv::GArg(in)});
+    auto check = [](const cv::GComputation &comp){
+        const auto &p = comp.priv();
+        EXPECT_EQ(3u, p.m_ins.size());
+        EXPECT_EQ(2u, p.m_outs.size());
 
-    EXPECT_EQ(1u, pattern.size());
-    EXPECT_EQ(1u, subst.size());
+        EXPECT_TRUE(ProtoContainsT<GMat>(p.m_ins[0]));
+        EXPECT_TRUE(ProtoContainsT<GScalar>(p.m_ins[1]));
+        EXPECT_TRUE(ProtoContainsT<cv::detail::GArrayU>(p.m_ins[2]));
+        EXPECT_TRUE(cv::util::get<cv::detail::GArrayU>(p.m_ins[2]).holds<int>());
+        EXPECT_FALSE(cv::util::get<cv::detail::GArrayU>(p.m_ins[2]).holds<char>());
 
-    auto checkOut = [](GArg& garg) {
-        EXPECT_TRUE(garg.kind == cv::detail::GTypeTraits<OutType>::kind);
-        EXPECT_NO_THROW(garg.get<OutType>());
+        EXPECT_TRUE(ProtoContainsT<GMat>(p.m_outs[0]));
+        EXPECT_TRUE(ProtoContainsT<GMat>(p.m_outs[1]));
     };
 
-    checkOut(pattern[0]);
-    checkOut(subst[0]);
+    check(tr.pattern());
+    check(tr.substitute());
 }
 
-TEST(KernelPackageTransform, GMat)
+TEST(KernelPackageTransform, gmat_in_garr_out)
 {
-    transformTest<gmat_in_gmat_out, GMat, GMat>();
-}
+    auto tr = gmat_in_garr_out::transformation();
 
-TEST(KernelPackageTransform, GMatP)
-{
-    transformTest<gmatp_in_gmatp_out, GMatP, GMatP>();
-}
+    auto check = [](const cv::GComputation &comp){
+        const auto &p = comp.priv();
+        EXPECT_EQ(1u, p.m_ins.size());
+        EXPECT_EQ(1u, p.m_outs.size());
 
-TEST(KernelPackageTransform, GScalarIn)
-{
-    transformTest<gsc_in_gmat_out, GScalar, GMat>();
-}
+        EXPECT_TRUE(ProtoContainsT<GMat>(p.m_ins[0]));
 
-TEST(KernelPackageTransform, GScalarOut)
-{
-    transformTest<gmat_in_gsc_out, GMat, GScalar>();
-}
+        EXPECT_TRUE(ProtoContainsT<cv::detail::GArrayU>(p.m_outs[0]));
+        EXPECT_TRUE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<int>());
+        EXPECT_FALSE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<float>());
+    };
 
-TEST(KernelPackageTransform, DISABLED_GArrayIn)
-{
-    transformTest<garr_in_gmat_out, GArray<int>, GMat>();
+    check(tr.pattern());
+    check(tr.substitute());
 }
 
-TEST(KernelPackageTransform, DISABLED_GArrayOut)
+TEST(KernelPackageTransform, garr_in_gmat_out)
 {
-    transformTest<gmat_in_garr_out, GMat, GArray<int>>();
```

Comment:
I propose to make a common test body for 1 input -> 1 output transformations (like `transformTest` was) and to test all possible in-out type combinations (check for GArray and be extended via template specialization of the check function)

Reply:
?